Updated unsupported node version error

patocallaghan opened the pull request:
Fixes ember-cli/ember-cli#5027

Removed `validPlatformVersion` utility function in favour of a new `PlatformChecker` module.

With the `PlatformChecker` we can now determine whether the Node platform is:

* `isValid` -  sits within the accepted version range
* `isDeprecated` - sits below the accepted version range
* `isUntested` - sits above the accepted version range

**Valid**
![image](https://cloud.githubusercontent.com/assets/685034/10905311/4a4a79c2-8211-11e5-9984-2ffa0511b35c.png)

**Untested**
![image](https://cloud.githubusercontent.com/assets/685034/10905302/37d4a16e-8211-11e5-9b50-72453c0ef7e1.png)

**Deprecated**
![image](https://cloud.githubusercontent.com/assets/685034/10905295/24a4581e-8211-11e5-8826-472d424aaa5e.png)

rwjblue commented:
This message looks good for Node v5.x, but we should still say that v0.10 is deprecated.

Can you tweak to ensure that the deprecation is printed for v0.10, and this newer (much nicer) warning is printed for other unknown versions?

patocallaghan commented:
@rwjblue so what does ember-cli officially support? I've noticed that we actually just [accept any node version that is greater than](https://github.com/ember-cli/ember-cli/blob/master/lib/utilities/valid-platform-version.js#L7) `>=0.12.0`. Should this be more granular? `>=0.12.0 <=5.0.0` perhaps?
If not, perhaps the only part of the deprecation error I need to change is the text about *Please update to Node 0.12 or io.js.*

rwjblue commented:
Sorry @patocallaghan, I got confused by how long it has been since we have released :crying_cat_face:. 

Looks like @stefanpenner and I are generally in agreement here though based on https://github.com/ember-cli/ember-cli/pull/4815#issuecomment-138739871.  I would like a deprecation for < 0.12 and a warning for > 4.0.0 (until we can test 5.x).

patocallaghan commented:
@rwjblue I've updated the PR and description above. I've introduced a new module for checking the platform and removed the previous lib that was used. Let me know if that is okay.

Looking at the `travis.yml` it looks like we run the [CI tests against v5](https://github.com/ember-cli/ember-cli/blob/master/.travis.yml#L8). Does this not imply that we already support Node v5?

Also should the warnings be yellow in both cases or red? Yellow implies a warning but should a deprecation message be red?

kellyselden commented on lib/utilities/platform-checker.js:
```diff
@@ -0,0 +1,28 @@
+var semver = require('semver');
+
+var LOWER_RANGE = '0.12.0';
+var UPPER_RANGE = '4.2.1';
```
This should be <5.0.0, because we want point releases like 4.2.2 to pass without a code change.

patocallaghan replied:
@kellyselden sure, makes sense. I've just updated it.

rwjblue commented:
> Looking at the travis.yml it looks like we run the CI tests against v5. Does this not imply that we already support Node v5?

Agreed. Let's make the upper bounds use 6.0.0.

patocallaghan commented:
@rwjblue I've updated to allow v5 in the valid range

rwjblue commented:
@homu r+

homu commented:
:pushpin: Commit 91262dc has been approved by `rwjblue`

<!-- @homu r=rwjblue 91262dce260f1375c9b9f301f38433af0fc6ce69 -->

homu commented:
:hourglass: Testing commit 91262dc with merge a92d5ca...

homu commented:
:sunny: Test successful - [status](https://travis-ci.org/ember-cli/ember-cli/builds/89011260)
